crates/sui-indexer-alt/migrations/2024-11-02-160143_kv_epochs/up.sql

```diff
@@ -0,0 +1,56 @@
+-- Information related to an epoch that is available when it starts
+CREATE TABLE IF NOT EXISTS kv_epoch_starts
+(
+    epoch                       BIGINT        PRIMARY KEY,
+    protocol_version            BIGINT        NOT NULL,
+
+    -- Inclusive checkpoint lowerbound of the epoch.
+    cp_lo                       BIGINT        NOT NULL,
+    -- The timestamp that the epoch starts at. This is always extracted from
+    -- the system state object.
+    start_timestamp_ms          BIGINT        NOT NULL,
+    -- The reference gas price that will be used for the rest of the epoch.
+    reference_gas_price         BIGINT        NOT NULL,
+    -- BCS serialized SystemState.
+    system_state                BYTEA         NOT NULL
+);
+
+-- Information related to an epoch that is available when it ends (after the
+-- epoch advancement to the next epoch)
+CREATE TABLE IF NOT EXISTS kv_epoch_ends
+(
+    epoch                       BIGINT        PRIMARY KEY,
+
+    -- Exclusive checkpoint upperbound of the epoch.
+    cp_hi                       BIGINT        NOT NULL,
+    -- Exclusive transaction upperbound of the epoch.
+    tx_hi                       BIGINT        NOT NULL,
```

Comment:
alternatively or additionally, at epoch boundary we also know the `tx_lo` because there must be at least 1 tx in the epoch

Reply:
I'm not sure I understand -- how do we figure out the tx_lo for the epoch without reading from the DB?